### ISSUE TITLE
Reflection

### ISSUE THREAD

**Issue by VegeTato** [GDevelopApp/GDevelop-extensions#186]
![Reflection](https://user-images.githubusercontent.com/81715237/125969707-30534d95-7cb3-41e9-8142-d704534d0670.gif)
Reflection extension is useful when it comes to reflecting objects
examples:
player walking on water/glass/beach

**You can test the extension here:**
https://games.gdevelop-app.com/game-3fa0644d-b7a7-4fe3-895f-fdb5a1ac3cc8/index.html

Version 0.1
-------------------------
-User can set the reflection opacity
-User can select which floor object to reflect on
-User can set the Z order of the reflection
-User have the option to turn ON/OFF to delete the reflection when it touch the original object
-User can set the touch distance to delete the reflection

Version 0.2
-------------------------
-Changed the action name to "Add reflection to object"
-Changed the distance label to "Distance offset"
-Width and Height of the original object is now taken into account for the reflection object
-User can create the reflection on a selected layer
-Changed the short description to "Add a reflection to a selected object"
-Changed the delete distance label to "Distance offset to delete reflection when it touch the original object (0 for immediate delete when touch)"

Version 0.3
-------------------------
-Reflection will follow the original object angle as a mirror
-Reflection will follow the visibility of the original object
-Reflection will follow the original object Current animation (Paused/resumed/animation number..etc)
-Changed reflection collide label to "Delete reflection when it collides with the original object"
-Changed reflection delete distance label to "Distance between the reflection and the original object before deleting the reflection (0 for immediate delete when touch)"

## Checklist
- [x] Extension has a proper name and description.
- [x] Extension has tags (for example: "platform, brick, breakable").
- [x] All behaviors have a description.
- [x] All functions (actions, conditions, expressions) have descriptions and proper sentences.
- [x] I've followed the best practices listed on http://wiki.compilgames.net/doku.php/gdevelop5/extensions/best-practices
- [x] I confirm that this extension can be integrated to this GitHub repository, distributed and MIT licensed.

**You can download the example here:**
https://trello-attachments.s3.amazonaws.com/5fd532f01e99c97130e99fe9/60e31e080537a84b026095ab/df74bdc5be6323d7d5d46cc6ba02bd81/Reflection_extension_example_V0.3.rar

**You can download the extension here:**
https://trello-attachments.s3.amazonaws.com/5fd532f01e99c97130e99fe9/60e31e080537a84b026095ab/38abcf33ccb28c411edb857d9a5efccb/Reflection_extension_V0.3.rar

**Comment by Midhil457** [comment]
Hi, great extension

- The action name could be "add reflection" as it is adding a reflection and not setting it. The sentence in event sheet is fine
- Isn't "Distance Offset" better describing than "Distance between the object and the reflection"? as it adding "distance" to the normal
- The reflection object should mimic all the visual properties of the original object. For Example: currently the size is not taken into account, opacity (maybe it should subtract the given opacity with the original object, or leave it for the user to customize through the parameters) and any visual property than can be changed
![Ref](https://user-images.githubusercontent.com/73597906/126036444-1d2905fb-ecff-4b21-89d3-0126fcc56562.PNG)

![Ref2](https://user-images.githubusercontent.com/73597906/126036445-e88e227f-23fa-400f-b5af-ce36170693dd.PNG)




- A parameter to choose the layer for the reflection object might be useful
- The label "Distance to delete reflection when it touch the original object" is a bit hard to understand
- The short description could be a bit more than just "Reflect object"
- I didn't really understand why there are conditions for setting the reflection's Y position. Won't `floor.Y()-(Object.Y()-floor.Y())-Object.Height()+GetArgumentAsNumber("distance")` work fine? it worked fine for me, the reflection has the equal distance to the floor as the original object.

:)

**Comment by VegeTato** [comment]
Thank you for your time to review my extension <3
>     * The action name could be "add reflection" as it is adding a reflection and not setting it. The sentence in event sheet is fine
>     * Isn't "Distance Offset" better describing than "Distance between the object and the reflection"? as it adding "distance" to the normal

i love your new names, ill make sure to rename my extension 👍 

> The reflection object should mimic all the visual properties of the original object. For Example: currently the size is not taken into account, opacity (maybe it should subtract the given opacity with the original object, or leave it for the user to customize through the parameters) and any visual property than can be changed

-i totally forgot about the object size, ill take that into account in the events. 👍 
-its better to leave the opacity to the user as its an optional thing.

>     * A parameter to choose the layer for the reflection object might be useful

Totally agree i need that option in my extension 👍 

>     * The label "Distance to delete reflection when it touch the original object" is a bit hard to understand

also agree, its long, hmmm ill try to come up with something if you don't have anything in mind :)

>     * The short description could be a bit more than just "Reflect object"

👍 

>     * I didn't really understand why there are conditions for setting the reflection's Y position. Won't `floor.Y()-(Object.Y()-floor.Y())-Object.Height()+GetArgumentAsNumber("distance")` work fine? it worked fine for me, the reflection has the equal distance to the floor as the original object.

now that's really weird, i think that was me testing few things and i left it like that, i totally agree it works exactly the same, with or without the conditions, thanks for bringing this to my attention !

ill fix everything mentioned here and replay it here and on trello and on discord, thank you for your review <3

**Comment by tristanbob** [comment]
> ill fix everything mentioned here and replay it here and on trello and on discord, thank you for your review <3

@VegeTato once an issue is started on Github, let's keep the discussion here (so you don't have to update both places).  You only need to update Trello to move it to the correct board when the status changes.

@Midhil457 thanks for your valuable work in reviewing this!

**Comment by VegeTato** [comment]
**Reflection extension updated to V0.2 based on the review <3**

-Changed the action name to "Add reflection to object"
-Changed the distance label to "Distance offset"
-Width and Height of the original object is now taken into account for the reflection object
-User can create the reflection on a selected layer
-Changed the short description to "Add a reflection to a selected object"
-Changed the delete distance label to "Distance offset to delete reflection when it touch the original object (0 for immediate delete when touch)"

**Comment by Midhil457** [comment]
Hi!
- Some of the visual changes still are not mimicked, Like angle, when hiding object, pause animation- 
![refl1](https://user-images.githubusercontent.com/73597906/126059500-eacd9094-d641-4c10-a47c-65eeef48841c.PNG)
![Refl2](https://user-images.githubusercontent.com/73597906/126059501-d64f50f0-7238-4292-98d3-4d59566736e1.PNG)
![Refl3](https://user-images.githubusercontent.com/73597906/126059503-f8b4e3f7-f57d-4faf-9366-9c6104eb34dc.PNG)

- `Delete reflection when it touch the original object ?` and `Distance offset to delete reflection when it touch the original object` still sounds confusing. I don't have much idea for this, but `Delete reflection when it colliding with the original object` and `Deletion distance padding` or `Margin distance between reflection and original after which the reflection should be deleted`. I am not sure about both

- ` The original object could be called `reference object`

:)

**Comment by VegeTato** [comment]
**Reflection extension updated to V0.3 based on the review** ❤️

-Reflection will follow the original object angle as a mirror
-Reflection will follow the visibility of the original object
-Reflection will follow the original object Current animation (Paused/resumed/animation number..etc)
-Changed reflection collide label to "Delete reflection when it collides with the original object"
-Changed reflection delete distance label to "Distance between the reflection and the original object before deleting the reflection (0 for immediate delete when touch)"

if the labels still confusing, well.. <.> i cant think of something better, you can suggest me :3
(original object is more friendly than reference object) i think 🤔

**Comment by Midhil457** [comment]
I think you forgot to update the example project, it still is version 0.2

**Comment by Midhil457** [comment]
Hi! I didn't find any issues and can be added (I don't know how to add it properly, so someone else would have to do it 😅 )

- I do have a suggestion, feel free to ignore it this is not necessary, but a horizontal reflection might be helpful and
also a "Delete when not colliding with floor (for something like mirror, etc....)

:)

**Comment by Bouh** [comment]
The result is really cool!
I'll review and publish this extension today,

**Comment by Bouh** [comment]
Ideally, this should be a behavior, but as the behaviors can't use more than the original object in its properties I release it with the actions.


I've reworked a bit the example and the extension for better support on the Zorder and the opacity, it looks better:

![image](https://user-images.githubusercontent.com/1670670/128516052-ae6908b1-6ddc-4df8-ade9-698192df70cc.png)


The reflection is under the player, but for moving it in front we have to mask the reflection object with another extension, this will be possible in the future with a feature called Behavior composition. (Using behaviors in another behavior).


The extension has been added into GDevelop.
Thanks guys @VegeTato , @Midhil457 !
